### PR TITLE
Enable caching to reduce RPC calls

### PR DESCRIPTION
**Pull request by PawlowskiAdrian**
enable caching, will reduce RPC calls for method chainID

### CODE REVIEW

**Comment by ErikBjare** [comment]
The simple_cache_middleware is probably too aggressive and will affect freshness of queried prices etc.

Perhaps look into some of the more sophisticated caching methods? Or we might just want to inform users how to do this themselves in the docs by constructing a custom web3 with a caching layer of choice.

**Comment by PawlowskiAdrian** [comment]
good call @ErikBjare , adding changes

**Comment by PawlowskiAdrian** [comment]
added changes, looking for any commentary

File: uniswap/uniswap.py [hunk 1]
```diff
@@ -115,7 +118,9 @@ def __init__(
                 provider = os.environ["PROVIDER"]
             self.w3 = Web3(Web3.HTTPProvider(provider, request_kwargs={"timeout": 60}))
 
-        # Cache netid to avoid extra RPC calls
+        if enable_caching == True:
```
Comment (ErikBjare) [inline uniswap/uniswap.py:121]:
```suggestion
        if enable_caching:
```

**Review by ErikBjare — COMMENTED** [review]
Looks good! Going to let CI run now and will merge if it works :)

File: uniswap/uniswap.py [hunk 2]
```diff
@@ -78,6 +79,7 @@ def __init__(
         # use_eip1559: bool = True,
         factory_contract_addr: str = None,
         router_contract_addr: str = None,
+        enable_caching: bool = False,
```
Comment (ErikBjare) [inline uniswap/uniswap.py:82]: Is this not safe to set to `True` now, given the whitelist?

Comment (PawlowskiAdrian) [reply uniswap/uniswap.py:82]: Should be safe, unless middleware adds some unintended behaviour. Haven't read that much about them. For all current users don't wanted to force them to run with middleware. I rather want them to read about middleware and set this flag manually.

Comment (ErikBjare) [reply uniswap/uniswap.py:82]: Agreed!

**Comment by ErikBjare** [comment]
Thanks for this @PawlowskiAdrian :)

**Pull request merged**